File: Data Structures & Algorithms/max-water-container/submission-2.java
class Solution {
    public int maxArea(int[] heights) {
         int l = 0, r = heights.length-1;
         int max = 0;
         while (l < r)
         {
            int h = Math.min(heights[l], heights[r]);
            int area = h * (r-l);
            max = Math.max(max,area);
            if (h == heights[l])
            {
                l += 1;
            } else {
                r -=1;
            }
         }
         return max;
    }
}

File: Data Structures & Algorithms/largest-rectangle-in-histogram/submission-2.cpp
class Solution {
public:
    int largestRectangleArea(vector<int> &heights)
    {
        int maxArea = 0;
        stack<pair<int, int>> st; // {i, h}
        for (int i = 0; i < heights.size(); i++)
        {
            int start = i;
            while (!st.empty() && st.top().second > heights[i])
            {
                int height = st.top().second;
                int index = st.top().first;
                maxArea = max(height * (i - index), maxArea);
                start = index;
                st.pop();
            }
            st.push({start, heights[i]});
        }
        while (!st.empty())
        {
            const auto &pair = st.top();
            int index = st.top().first;
            int height = st.top().second;
            maxArea = max(maxArea, (height * (static_cast<int>(heights.size()) - index)));
            st.pop();
        }

        return maxArea;
    }


};
